src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const copy = { ...currentState };
+        action.keysToRemove.forEach((key) => {
+          delete copy[key];
+        });
+
+        currentState = { ...copy };
+        break;
+
+      default:
+        break;
+    }
+
+    statesArray.push(currentState);
```

Comment:
You are pushing the `currentState` reference into `statesArray`. This means all elements in `statesArray` will reference the same object, which is not the desired behavior. Instead, you should push a copy of `currentState` to ensure each element in the array is a distinct object.